File: myFeed.Tests/Fixtures/ParallelFeedStoreTests.fs
module myFeed.Tests.Fixtures.ParallelFeedStoreTests

open Xunit
open NSubstitute
open myFeed.Services.Abstractions
open myFeed.Services.Implementations
open myFeed.Services.Models
open myFeed.Tests.Extensions
open System.Threading.Tasks
open System

[<Fact>]
let ``should sort stored article entities``() = 

    let settings = Substitute.For<ISettingService>()
    settings.GetAsync<_>(Arg.Any()).Returns(5) |> ignore

    let fetcher = Substitute.For<IFeedFetchService>()
    fetcher.FetchAsync(Arg.Any<_>()).Returns(
        Seq.empty<Article> 
        |> Task.FromResult)
        |> ignore

    let service = produce<ParallelFeedStoreService> [fetcher; settings]
    let articles =
        service.LoadAsync(
            [ Channel(Articles=toList
                [ Article(Title="Foo", PublishedDate=DateTime.Now);
                  Article(Title="Bar", PublishedDate=DateTime.MinValue);
                  Article(Title="Zoo", PublishedDate=DateTime.MaxValue) ]
              )]).Result
        |> List.ofSeq              

    Should.equal 3 articles.Length
    Should.equal "Zoo" articles.[0].Title
    Should.equal "Foo" articles.[1].Title
    Should.equal "Bar" articles.[2].Title

[<Fact>]
let ``should save fetched article entities``() =    

    let settings = Substitute.For<ISettingService>()
    settings.GetAsync<_>(Arg.Any()).Returns(5) |> ignore

    let fetcher = Substitute.For<IFeedFetchService>()
    fetcher.FetchAsync(Arg.Any<_>()).Returns(
        [Article(Title="Foo")] :> seq<_>
        |> Task.FromResult)
        |> ignore

    let mutable articlesInserted = null
    let categories = Substitute.For<ICategoryStoreService>()
    categories.When(fun x -> x.InsertArticleRangeAsync(Arg.Any<_>(), Arg.Any<_>()) |> ignore)
              .Do(fun x -> articlesInserted <- x.Arg<seq<Article>>())       

    let service = produce<ParallelFeedStoreService> [fetcher; categories; settings]
    let articles = 
        service.LoadAsync(
            [ Channel(Uri="http://foo.bar") ]).Result
            |> List.ofSeq            

    let article = Seq.item 0 articlesInserted
    Should.equal 1 articles.Length
    Should.equal "Foo" articles.[0].Title
    Should.equal "Foo" article.Title

[<Fact>]
let ``should mix and order fetched and stored articles by date``() = 
    
    let settings = Substitute.For<ISettingService>()
    settings.GetAsync<_>(Arg.Any()).Returns(5) |> ignore

    let fetcher = Substitute.For<IFeedFetchService>()
    fetcher.FetchAsync(Arg.Any<_>()).Returns(
        [Article(Title="Foo", PublishedDate=DateTime.Now)] :> seq<_>
        |> Task.FromResult)
        |> ignore

    let service = produce<ParallelFeedStoreService> [fetcher; settings]
    let articles =
        service.LoadAsync(
            [ Channel(Articles=toList
                [ Article(Title="Bar", PublishedDate=
                    DateTime.MinValue)])]).Result
        |> List.ofSeq

    Should.equal 2 articles.Length
    Should.equal "Foo" articles.[0].Title
    Should.equal "Bar" articles.[1].Title         

[<Fact>]
let ``should remove outdated articles if count is greater than custom``() = 

    let settings = Substitute.For<ISettingService>()
    settings.GetAsync<_>(Arg.Any()).Returns(70) |> ignore

    let articles = Seq.init 200 (fun _ -> Article())
    let channel = Channel(Articles=toList articles)
    let fetcher = Substitute.For<IFeedFetchService>()
    fetcher.FetchAsync(Arg.Any()).Returns(Seq.empty 
        |> Task.FromResult) |> ignore
    
    let service = produce<ParallelFeedStoreService> [fetcher; settings]
    let articles = service.LoadAsync([channel]).Result |> List.ofSeq

    Should.equal 70 articles.Length

[<Fact>]
let ``should remove articles with minimum publishing date only``() =    

    let settings = Substitute.For<ISettingService>()
    settings.GetAsync<_>(Arg.Any()).Returns(70) |> ignore

    let articles = Seq.init 200 (fun _ -> Article())
    let fetcher = Substitute.For<IFeedFetchService>()
    fetcher.FetchAsync(Arg.Any()).Returns(articles 
        |> Task.FromResult) |> ignore
    
    let service = produce<ParallelFeedStoreService> [fetcher; settings]
    let articles = service.LoadAsync([Channel()]).Result |> List.ofSeq

    Should.equal 70 articles.Length
    
[<Fact>]
let ``should ignore whitespaces while comparing titles``() =
    
    let settings = Substitute.For<ISettingService>()
    settings.GetAsync<_>(Arg.Any()).Returns(70) |> ignore
    
    let fetcher = Substitute.For<IFeedFetchService>()
    fetcher.FetchAsync(Arg.Any<_>()).Returns(
        [ Article(Title="Foo  ", FeedTitle="Bar");
          Article(Title="Bar\r\n", FeedTitle="Foo")] :> seq<_>
        |> Task.FromResult) |> ignore
    
    let service = produce<ParallelFeedStoreService> [fetcher; settings]
    let articles = 
        service.LoadAsync(
            [ Channel(Articles=toList 
                [ Article(Title="Foo", FeedTitle="Bar");
                  Article(Title="Bar", FeedTitle="Foo") ]) ]).Result 
                |> List.ofSeq
    
    Should.equal 2 articles.Length
    Should.equal "Foo" articles.[0].Title
    Should.equal "Bar" articles.[1].Title
    Should.equal "Bar" articles.[0].FeedTitle
    Should.equal "Foo" articles.[1].FeedTitle

File: MyFeed.Tests/ParallelFeedStoreTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using FluentAssertions;
using MyFeed.Interfaces;
using MyFeed.Models;
using MyFeed.Services;
using NSubstitute;
using Xunit;

namespace MyFeed.Tests
{
    public sealed class ParallelFeedStoreTests
    {
        private readonly IFeedFetchService _feedFetchService = Substitute.For<IFeedFetchService>();
        private readonly ICategoryManager _categoryManager = Substitute.For<ICategoryManager>();
        private readonly ISettingManager _settingManager = Substitute.For<ISettingManager>();
        private readonly IFeedStoreService _feedStoreService;

        public ParallelFeedStoreTests() => _feedStoreService = new ParallelFeedStoreService(
            _feedFetchService, _categoryManager, _settingManager
        );
        
        [Fact]
        public async Task ShouldSortStoredArticleEntities()
        {
            _settingManager.Read().Returns(new Settings {Max = 5});
            _feedFetchService.Fetch(Arg.Any<string>()).Returns(new List<Article>());
            var channel = new Channel {Articles = new List<Article>
            {
                new Article {Title = "Foo", PublishedDate = DateTime.Now},
                new Article {Title = "Bar", PublishedDate = DateTime.MinValue},
                new Article {Title = "Zoo", PublishedDate = DateTime.MaxValue}
            }};
            
            var result = await _feedStoreService.Load(new List<Channel> {channel});
            var articles = new List<Article>(result);

            articles.Count.Should().Be(3);
            articles[0].Title.Should().Be("Zoo");
            articles[1].Title.Should().Be("Foo");
            articles[2].Title.Should().Be("Bar");
        }

        [Fact]
        public async Task ShouldSaveFetchedArticleEntities()
        {
            _settingManager.Read().Returns(new Settings {Max = 5});
            _feedFetchService.Fetch(Arg.Any<string>()).Returns(new List<Article>{new Article {Title = "Foo"}});
            var channels = new List<Channel> {new Channel {Uri = "http://foo.bar"}};

            IEnumerable<Article> inserted = null;
            _categoryManager
                .When(x => x.Update(Arg.Any<Channel>()))
                .Do(callback => inserted = callback.Arg<Channel>().Articles);
            
            var result = await _feedStoreService.Load(channels);

            var articlesInserted = new List<Article>(inserted);
            articlesInserted[0].Title.Should().Be("Foo");
            articlesInserted.Count.Should().Be(1);
            
            var articlesLoaded = new List<Article>(result);
            articlesLoaded[0].Title.Should().Be("Foo");
            articlesLoaded.Count.Should().Be(1);
        }

        [Fact]
        public async Task ShouldOrderFetchedAndStoredArticlesByDate()
        {
            var fetchedArticles = new List<Article> {new Article {Title = "Foo", PublishedDate = DateTime.Now}};
            _feedFetchService.Fetch(Arg.Any<string>()).Returns(fetchedArticles);
            _settingManager.Read().Returns(new Settings {Max = 5});
            
            var result = await _feedStoreService.Load(new List<Channel> {new Channel
            {
                Articles = new List<Article> {new Article {Title = "Bar", PublishedDate = DateTime.MinValue}}
            }});
            
            var articles = new List<Article>(result);
            articles[0].Title.Should().Be("Foo");
            articles[1].Title.Should().Be("Bar");
            articles.Count.Should().Be(2);
        }

        [Fact]
        public async Task ShouldRemoveOutdatedArticlesIfCountIsCreaterThenLimit()
        {
            _settingManager.Read().Returns(new Settings {Max = 70});
            _feedFetchService.Fetch(Arg.Any<string>()).Returns(new List<Article>());
            var result = await _feedStoreService.Load(new List<Channel> {new Channel
            {
                Articles = Enumerable.Repeat(new Article(), 200).ToList()
            }});
            var articles = new List<Article>(result);
            articles.Count.Should().Be(70);
        }

        [Fact]
        public async Task ShouldIgnoreWhitespaceWhenComparingArticles()
        {
            _settingManager.Read().Returns(new Settings {Max = 5});
            _feedFetchService.Fetch(Arg.Any<string>()).Returns(new List<Article>
            {
                new Article {Title = " Foo   ", FeedTitle = "Bar"},
                new Article {Title = "Bar\r\n", FeedTitle = "Foo"}
            });
            var result = await _feedStoreService.Load(new List<Channel> {new Channel {Articles = new List<Article>
            {
                new Article {Title = "Foo", FeedTitle = "Bar"},
                new Article {Title = "Bar", FeedTitle = "Foo"}
            }}});

            var articles = new List<Article>(result);
            articles.Count.Should().Be(2);
            articles.Last().Title.Should().Be("Bar");
            articles.First().Title.Should().Be("Foo");
            articles.First().FeedTitle.Should().Be("Bar");
            articles.Last().FeedTitle.Should().Be("Foo");
        }
    }
}